### PR TITLE
feat(topology): Add topology to dsp and migrate to v1beta2

### PR DESCRIPTION
**Pull request by sinhaashish**
This PR adds the topology to DSP and bumps up the version to v1beta2. 
The code for migration of old dsp is also a part of it.

### CODE REVIEW

File: k8s/operators/src/pool/diskpool/migration.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    api::{Patch, PatchParams, PostParams},
+    core::crd::merge_crds,
+    Api, Client, CustomResourceExt, ResourceExt,
+};
+use std::time::Duration;
+use tracing::{error, info};
+
+const PAGINATION_LIMIT: u32 = 100;
+
+/// In case of v1alpha1 and v1beta1 check, ensure that crd exist and tehn migrate to v1beta2.
```
Comment (abhilashshetty04) [inline k8s/operators/src/pool/diskpool/migration.rs]: then*

File: k8s/operators/src/pool/diskpool/migration.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+///
+/// To prevent that, we will simply panic, and hope we can make progress after restart. Keep
+/// running is not an option as the operator would be "running" and the only way to know something
+/// is wrong would be to consult the logs.
+pub(crate) async fn ensure_crd(
+    k8s: &Client,
+    api_version: &ApiVersion,
+) -> Result<CustomResourceDefinition, Error> {
+    let crd_api: Api<CustomResourceDefinition> = Api::all(k8s.clone());
+    // Check if there is an existing dsp diskpool. If yes, Replace it with new generated one.
```
Comment (abhilashshetty04) [inline k8s/operators/src/pool/diskpool/migration.rs]: In the immediate next line you are just creating existing CRD right? Not replacing.

**Review by abhilashshetty04 — APPROVED** [review]

File: k8s/operators/src/pool/diskpool/v1beta1.rs [hunk 3]
```diff
@@ -35,14 +35,17 @@ pub struct DiskPoolSpec {
 
 impl DiskPoolSpec {
     /// Create a new DiskPoolSpec from the node and the disks.
+    #[allow(dead_code)]
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/diskpool/v1beta1.rs:38]: if not used anymore please remove.

File: k8s/operators/src/pool/main.rs [hunk 4]
```diff
@@ -89,50 +89,41 @@ async fn reconcile(dsp: Arc<DiskPool>, ctx: Arc<OperatorContext>) -> Result<Acti
...
 
+/// Api version of DSP.
+#[derive(Debug, Clone, PartialEq)]
+pub enum ApiVersion {
+    /// Represents v1alpha1
+    Alpha1,
+    /// Represents v1beta1
+    Beta1,
+    /// Represents v1beta2
+    Beta2,
```
Comment (tiagolobocastro) [inline k8s/operators/src/pool/main.rs:100]:
```suggestion
    V1Alpha1,
    /// Represents v1beta1
    V1Beta1,
    /// Represents v1beta2
    V1Beta2,
```

**Review by tiagolobocastro — APPROVED** [review]
suggest additional refactor if you want to take up, add crd folder and move v1beta* along with migration there.

**Comment by sinhaashish** [comment]
> // Check if there is an existing dsp diskpool.



> suggest additional refactor if you want to take up, add crd folder and move v1beta* along with migration there.

will take this up in another PR

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Comment by sinhaashish** [comment]
bors merge

**Pull request merged**